### PR TITLE
Backport "modify()" method from latest mongoengine

### PR DESCRIPTION
**Pull request by thomasst**
This PR backports the `modify()` method from latest mongoengine.

@closeio/engineering / @wojcikstefan please review.

### CODE REVIEW

File: mongoengine/queryset/queryset.py [hunk 1]
```diff
@@ -543,6 +543,60 @@ def update_one(self, upsert=False, write_concern=None, **update):
...
+        if remove and new:
+            raise OperationError("Conflicting parameters: remove and new")
+
+        if not update and not upsert and not remove:
+            raise OperationError(
+                "No update parameters, must either update or remove")
+
+        queryset = self.clone()
+        query = queryset._query
+        update = transform.update(queryset._document, **update)
```
Comment (wojcikstefan) [inline mongoengine/queryset/queryset.py]: https://github.com/MongoEngine/mongoengine/blob/master/mongoengine/queryset/base.py#L559 has a conditional: `if not IS_PYMONGO_3 or not remove:`. Should this happen if `remove` is true for us? Also, we eventually want to support pymongo 3+ (mongodb v3.2 requires it). If so, should we backport it in a way where we won't have to backport it again once we make the rest of the codebase compatible w/ pymongo v3.2?

File: mongoengine/queryset/queryset.py [hunk 2]
```diff
@@ -543,6 +543,60 @@ def update_one(self, upsert=False, write_concern=None, **update):
...
+            raise NotUniqueError(u"Update failed (%s)" % err)
+        except pymongo.errors.OperationFailure, err:
+            raise OperationError(u"Update failed (%s)" % err)
+
+        if full_response:
+            if result["value"] is not None:
+                result["value"] = self._document._from_son(result["value"])
+        else:
+            if result is not None:
+                result = self._document._from_son(result)
```
Comment (wojcikstefan) [inline mongoengine/queryset/queryset.py]: Is there a reason why we dropped `only_fields` here? The upstream method does `self._document._from_son(result, only_fields=self.only_fields)` here.

Comment (thomasst) [inline mongoengine/queryset/queryset.py]: We don't have that in our mongoengine version -- they made changes to how "only" fields are handled.

File: mongoengine/queryset/queryset.py (hunk 1 above)
Comment (thomasst) [inline mongoengine/queryset/queryset.py]: `not IS_PYMONGO_3` is true in our case, and `True or X` is therefore always true.

There are several places where they use `not IS_PYMONGO_3`. This is just a backport to our version which doesn't technically support pymongo3, and I also don't have it in my dev environment to test right now. There are more variables / imports that need to be backported for full pymongo3 support.

Once we're ready for pymongo3 we should do a separate review and testing.

**Pull request merged**